L1Trigger/TrackerTFP/interface/DataFormats.h

```diff
@@ -0,0 +1,672 @@
+    // helper (loop) to convert bits to ntuple of variables
+    template<int it = 0, typename ...Ts>
+    void extractTrack(TTBV& ttBV, std::tuple<Ts...>& data, Process p) const;
+    // helper (loop) to convert ntuple of variables to bits
+    template<int it = 0, typename... Ts>
+    void attachStub(const std::tuple<Ts...>& data, TTBV& ttBV, Process p) const;
+    // helper (loop) to convert ntuple of variables to bits
+    template<int it = 0, typename... Ts>
+    void attachTrack(const std::tuple<Ts...>& data, TTBV& ttBV, Process p) const;
+    // configuration during construction
+    edm::ParameterSet iConfig_;
+    // stored run-time constants
+    const trackerDTC::Setup* setup_;
+    // collection of unique formats
+    std::vector<DataFormat> dataFormats_;
+    // variable flavour mapping
+    std::vector<std::vector<DataFormat*>> formats_;
+    // number of unused frame bits for a all Stub flavours
+    std::vector<int> numUnusedBitsStubs_;
+    // number of unused frame bits for a all Track flavours
+    std::vector<int> numUnusedBitsTracks_;
+    // number of channels of all processes on a TFP
+    std::vector<int> numChannel_;
+    // number of channels of all processes for whole system
+    std::vector<int> numStreams_;
+  };
+
+  // base class to represent stubs
+  template<typename ...Ts>
+  class Stub {
+  public:
+    // construct Stub from Frame
+    Stub(const TTDTC::Frame& frame, const DataFormats* dataFormats, Process p);
+    template<typename ...Others>
+    // construct Stub from other Stub
+    Stub(const Stub<Others...>& stub, Ts... data);
+    // construct Stub from TTStubRef
+    Stub(const TTStubRef& ttStubRef, const DataFormats* dataFormats, Process p, Ts... data);
+    Stub() {}
+    ~Stub() {}
+    // true if frame valid, false if gap in data stream
+    explicit operator bool() const { return frame_.first.isNonnull(); }
+    // access to DataFormats
+    const DataFormats* dataFormats() const { return dataFormats_; }
+    // stub flavour
+    Process p() const { return p_; }
+    // acess to frame
+    const TTDTC::Frame& frame() const { return frame_; }
+    // access to TTStubRef
+    const TTStubRef& ttStubRef() const { return frame_.first; }
+    // access to bitvector
+    const TTDTC::BV& bv() const { return frame_.second; }
+    // id of collection this stub belongs to
+    int trackId() const { return trackId_; }
+  protected:
+    // number of used bits for given variable
+    int width(Variable v) const { return dataFormats_->width(v, p_); }
+    // precision of given variable
+    double base(Variable v) const { return dataFormats_->base(v, p_); }
+    // format of given variable
+    const DataFormat& format(Variable v) const { return dataFormats_->format(v, p_); }
+    // all dataformats
+    const DataFormats* dataFormats_;
+    // stub flavour
+    Process p_;
+    // underlying frame
+    TTDTC::Frame frame_;
+    // ntuple of variables this stub is assemled of
+    std::tuple<Ts...> data_;
```

Comment:
I think the stub variables are stored twice in this class, once as a 64b word in the frame_ and once as a tuple of variables in data_. I assume this is to save CPU converting between the two.

Reply:
done

Reply:
Question in original comment not answered?

Reply:
oh sorry the answer is yes.

Reply:
This is one of the most difficult classes to understand. Perhaps add to comment at start of class the comment "In data members of classes Stub* & Track* below, the variables describing stubs/tracks are stored both in digitial format as a 64b word in frame_, and in undigitized format in an std::tuple. (This saves CPU)".

As your code makes heavy use of frame_ & tuple, it's worth spelling it out.